add options for windows pack command.

BryanAldrich opened the pull request:
- ability to extract the bundle from a setup
- ability to create a setup from a bundle
- refactored the WindowsPackCommand for SetupPackage to be static to reduce the qty of new code
- added the ability to have sub commands defined by the commands instead of nested in the Program.cs (needed to pass the IServiceProvider around).

Commit:
add options for windows pack command.
ability to extract the bundle from a setup
ability to create a setup from a bundle
refactored the WindowsPackCommand for SetupPackage to be static to reduce the qty of new code
added the ability to have sub commands defined by the commands instead of nested in the Program.cs (needed to pass the IServiceProvider around).

BryanAldrich commented:
I added these so I would only have to store a single file in our artifacts CI/CD pipeline. On final deployment, the 'other' file can be made.

caesay commented:
P.S. I love your enthusiasm, thanks for all your PR's. If you'd like to discuss any changes before spending time on them, feel free to join our Discord and @ me! 

It is technically possible to extract the `nupkg` from a Windows setup, but it is not possible to do that on Mac and we need to consider carefully before adding os-specific features. I also can't think of any realistic use-cases for doing this. The `-full.nupkg` package is "the source of truth". You really should be keeping those around longer than any other types of assets. Therefore you would not need to extract it from Setup.exe. NOTE: the .nupkg you extract from a Setup.exe is **NOT** a "portable" package (eg. the `-Portable.zip` you get). The portable zip is constructed differently than the .nupkg zip is and contains items that are not available in the .nupkg, so the naming of the commands and text is not quite right.

For the other way around, could make sense to make a command available which converts a .nupkg into a setup, but it should be cross platform and at least support windows and mac. It probably shouldn't be nested inside the `pack` command (the delta commands are not nested either, for example). The complexity here, as you've probably already spotted, is codesigning. We'd be duplicating a significant number of options (all the available signing for both windows & mac at least). Also creating the possibility of code signing failing the second time around, or even using different signing certificates? I don't think Apple would be happy with that.

BryanAldrich commented:
This commit started because I wanted to store just one file in our CI pipeline, and then generate the delta, full package and the releases file when it was actually time for release. 90% of our builds never get promoted, so i really needed it to be dynamic.

I started with the "extract the nupkg from the bundle", but I ran into issues with how it was made (before i found the bundle helper methods). I finished the create setup bundle from nupkg, then went back and finished out the create nupkg from setup.


That being said, you're right, I don't have to deal with mac, we're a windows only shop and our app distribution is tiny and a fixed environment.

Without regard for ensuring that the users did actually use the same options every time, we could use interfaces on the options to ensure that the same ones were presented to the user every time. I did see all the extra codesign code, hence the static methods and I passed around the existing windows pack options rather than make something that took 7 parameters. I'm all for less code maintenance. (ps, i think that maintenance is a four letter word, gotta do it, but i hate it).

Commit:
rename ExtractPortable to ExtractNuget
Add interfaces for IWindowsSigningOptions to ensure the same options are available (at least in code).
Add stubs for IOsxSetupPackageOptions (incomplete).

caesay commented:
- [ ]
